# FRED API Categories

In [3]:
from typing import Any
from lib.mcp_client import MCPClient, MCPClientConfig
from lib.utils import print_json_vertical

MCP_URL = 'http://localhost:8080/sse'

In [5]:
config = MCPClientConfig(url=MCP_URL)

async def call_tool(tool_name: str, arguments: dict[str, Any] | None = None):
    async with MCPClient(config) as client:
        result = await client.call_tool(tool_name, arguments or {})
        print_json_vertical(result.model_dump())


async def explore_categories(root_id: int = 0, depth: int = 2):
    async with MCPClient(config) as client:
        queue = [(root_id, 0)]
        while queue:
            category_id, level = queue.pop(0)
            indent = "  " * level
            print(f"{indent}- category {category_id}")
            if level >= depth:
                continue
            response = await client.call_tool("fred.category_children", {"category_id": category_id})
            payload = response.structuredContent or {}
            for child in payload.get("categories", []):
                queue.append((child["id"], level + 1))

tool_name = "fred.category_children"

## First Level

In [6]:
args = {"category_id": 0}
result = await call_tool(tool_name, args)


{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 32991,\n      \"name\": \"Money, Banking, & Finance\",\n      \"parent_id\": 0\n    },\n    {\n      \"id\": 10,\n      \"name\": \"Population, Employment, & Labor Markets\",\n      \"parent_id\": 0\n    },\n    {\n      \"id\": 32992,\n      \"name\": \"National Accounts\",\n      \"parent_id\": 0\n    },\n    {\n      \"id\": 1,\n      \"name\": \"Production & Business Activity\",\n      \"parent_id\": 0\n    },\n    {\n      \"id\": 32455,\n      \"name\": \"Prices\",\n      \"parent_id\": 0\n    },\n    {\n      \"id\": 32263,\n      \"name\": \"International Data\",\n      \"parent_id\": 0\n    },\n    {\n      \"id\": 3008,\n      \"name\": \"U.S. Regional Data\",\n      \"parent_id\": 0\n    },\n    {\n      \"id\": 33060,\n      \"name\": \"Academic Data\",\n      \"parent_id\": 0\n    }\n  ]\n}",
      "type": "text"
    }
  ],
  "isError": false,
 

## Money, Banking, & Finance

cat_id = 32991

In [7]:
args = {"category_id": 32991}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 22,\n      \"name\": \"Interest Rates\",\n      \"parent_id\": 32991\n    },\n    {\n      \"id\": 15,\n      \"name\": \"Exchange Rates\",\n      \"parent_id\": 32991\n    },\n    {\n      \"id\": 24,\n      \"name\": \"Monetary Data\",\n      \"parent_id\": 32991\n    },\n    {\n      \"id\": 46,\n      \"name\": \"Financial Indicators\",\n      \"parent_id\": 32991\n    },\n    {\n      \"id\": 23,\n      \"name\": \"Banking\",\n      \"parent_id\": 32991\n    },\n    {\n      \"id\": 32360,\n      \"name\": \"Business Lending\",\n      \"parent_id\": 32991\n    },\n    {\n      \"id\": 32145,\n      \"name\": \"Foreign Exchange Intervention\",\n      \"parent_id\": 32991,\n      \"notes\": \"In addition to the listed daily intervention series, intra-day series from the Swiss National Bank are available at: \\r\\nhttp://research.stlouisfed.org/fei/\\r\\n\

### Interest Rates

cat_id = 22

In [14]:
args = {"category_id": 22}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 34009,\n      \"name\": \"AMERIBOR Benchmark Rates\",\n      \"parent_id\": 22\n    },\n    {\n      \"id\": 33058,\n      \"name\": \"Automobile Loan Rates\",\n      \"parent_id\": 22\n    },\n    {\n      \"id\": 51,\n      \"name\": \"Bankers Acceptance Rate\",\n      \"parent_id\": 22\n    },\n    {\n      \"id\": 121,\n      \"name\": \"Certificates of Deposit\",\n      \"parent_id\": 22\n    },\n    {\n      \"id\": 120,\n      \"name\": \"Commercial Paper\",\n      \"parent_id\": 22\n    },\n    {\n      \"id\": 32348,\n      \"name\": \"Corporate Bonds\",\n      \"parent_id\": 22\n    },\n    {\n      \"id\": 33059,\n      \"name\": \"Credit Card Loan Rates\",\n      \"parent_id\": 22\n    },\n    {\n      \"id\": 34005,\n      \"name\": \"EONIA Rates\",\n      \"parent_id\": 22\n    },\n    {\n      \"id\": 34007,\n      \"name\": \"Euro Short-Term 

#### AMERIBOR Benchmark Rates

cat_id = 34009

In [15]:
args = {"category_id": 34009}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": []\n}",
      "type": "text"
    }
  ],
  "isError": false,
  "meta": null,
  "structuredContent": {
    "result": {
      "categories": []
    }
  }
}


#### Automobile Loan Rates

cat_id = 33058

#### Automobile Loan Rates

cat_id = 33058

## Population, Employment, & Labor Markets

cat_id = 10

In [8]:
args = {"category_id": 10}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 12,\n      \"name\": \"Current Population Survey (Household Survey)\",\n      \"parent_id\": 10,\n      \"notes\": \"The Current Population Survey (Household Survey) is a monthly survey of households conducted by the Bureau of Labor Statistics. It provides a comprehensive body of data on the labor force, employment, unemployment, persons not in the labor force, hours of work, earnings, and other demographic and labor force characteristics. \\r\\nNumerous conceptual and methodological differences between the current population (household) and establishment surveys result in important distinctions in the employment estimates derived from the surveys. Among these are: \\r\\nThe household survey includes agricultural workers, the self- employed, unpaid family workers, and private household workers among the employed. These groups are excluded from the establishm

## National Accounts

cat_id = 32992

In [9]:
args = {"category_id": 32992}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 18,\n      \"name\": \"National Income & Product Accounts\",\n      \"parent_id\": 32992\n    },\n    {\n      \"id\": 5,\n      \"name\": \"Federal Government Debt\",\n      \"parent_id\": 32992\n    },\n    {\n      \"id\": 32251,\n      \"name\": \"Flow of Funds\",\n      \"parent_id\": 32992\n    },\n    {\n      \"id\": 13,\n      \"name\": \"U.S. Trade & International Transactions\",\n      \"parent_id\": 32992\n    }\n  ]\n}",
      "type": "text"
    }
  ],
  "isError": false,
  "meta": null,
  "structuredContent": {
    "result": {
      "categories": [
        {
          "id": 18,
          "name": "National Income & Product Accounts",
          "parent_id": 32992
        },
        {
          "id": 5,
          "name": "Federal Government Debt",
          "parent_id": 32992
        },
        {
          "id": 32251,
          "name": "Flow of F

## Production & Business Activity

cat_id = 1

In [10]:
args = {"category_id": 1}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 32262,\n      \"name\": \"Business Cycle Expansions & Contractions\",\n      \"parent_id\": 1\n    },\n    {\n      \"id\": 33936,\n      \"name\": \"Business Surveys\",\n      \"parent_id\": 1\n    },\n    {\n      \"id\": 32436,\n      \"name\": \"Construction\",\n      \"parent_id\": 1\n    },\n    {\n      \"id\": 33940,\n      \"name\": \"Emissions\",\n      \"parent_id\": 1\n    },\n    {\n      \"id\": 33955,\n      \"name\": \"Expenditures\",\n      \"parent_id\": 1\n    },\n    {\n      \"id\": 33490,\n      \"name\": \"Finance Companies\",\n      \"parent_id\": 1\n    },\n    {\n      \"id\": 32216,\n      \"name\": \"Health Insurance\",\n      \"parent_id\": 1\n    },\n    {\n      \"id\": 97,\n      \"name\": \"Housing\",\n      \"parent_id\": 1\n    },\n    {\n      \"id\": 3,\n      \"name\": \"Industrial Production & Capacity Utilization\",\n 

## International Data

cat_id = 32263

In [11]:
args = {"category_id": 32263}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 32264,\n      \"name\": \"Countries\",\n      \"parent_id\": 32263\n    },\n    {\n      \"id\": 32955,\n      \"name\": \"Geography\",\n      \"parent_id\": 32263,\n      \"notes\": \"This category contains both data aggregated for a specific geographic region and a way to find country categories based upon geographic region.\"\n    },\n    {\n      \"id\": 32265,\n      \"name\": \"Indicators\",\n      \"parent_id\": 32263\n    },\n    {\n      \"id\": 32956,\n      \"name\": \"Institutions\",\n      \"parent_id\": 32263,\n      \"notes\": \"This category contains data aggregated for multinational institutions and a way to find country series by membership to institutions.\"\n    }\n  ]\n}",
      "type": "text"
    }
  ],
  "isError": false,
  "meta": null,
  "structuredContent": {
    "result": {
      "categories": [
        {
          "id": 32264,
   

## U.S. Regional Data

cat_id = 3008

In [12]:
args = {"category_id":  3008}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 27281,\n      \"name\": \"States\",\n      \"parent_id\": 3008\n    },\n    {\n      \"id\": 32043,\n      \"name\": \"Census Regions\",\n      \"parent_id\": 3008\n    },\n    {\n      \"id\": 32061,\n      \"name\": \"BEA Regions\",\n      \"parent_id\": 3008\n    },\n    {\n      \"id\": 32849,\n      \"name\": \"BLS Regions\",\n      \"parent_id\": 3008\n    },\n    {\n      \"id\": 32071,\n      \"name\": \"Federal Reserve Districts\",\n      \"parent_id\": 3008\n    },\n    {\n      \"id\": 32233,\n      \"name\": \"Freddie Mac Regions\",\n      \"parent_id\": 3008\n    }\n  ]\n}",
      "type": "text"
    }
  ],
  "isError": false,
  "meta": null,
  "structuredContent": {
    "result": {
      "categories": [
        {
          "id": 27281,
          "name": "States",
          "parent_id": 3008
        },
        {
          "id": 32043,
          "

## Academic Data

cat_id = 33060

In [13]:
args = {"category_id": 33060}
result = await call_tool(tool_name, args)

{
  "content": [
    {
      "annotations": null,
      "meta": null,
      "text": "{\n  \"categories\": [\n    {\n      \"id\": 33833,\n      \"name\": \"Banking and Monetary Statistics, 1914-1941\",\n      \"parent_id\": 33060,\n      \"notes\": \"This data is from the Board of Governors' Banking and Monetary Statistics 1914-1941 publication.\"\n    },\n    {\n      \"id\": 33951,\n      \"name\": \"Daily Federal Funds Rate, 1928-54\",\n      \"parent_id\": 33060\n    },\n    {\n      \"id\": 33825,\n      \"name\": \"Data on the nominal term structure model from Kim and Wright\",\n      \"parent_id\": 33060\n    },\n    {\n      \"id\": 33891,\n      \"name\": \"Historical Federal Reserve Data\",\n      \"parent_id\": 33060\n    },\n    {\n      \"id\": 33061,\n      \"name\": \"NBER Macrohistory Database\",\n      \"parent_id\": 33060\n    },\n    {\n      \"id\": 33100,\n      \"name\": \"Penn World Table 7.1\",\n      \"parent_id\": 33060\n    },\n    {\n      \"id\": 33402,\n  